datastore/src/main/java/org/jboss/pnc/datastore/DefaultDatastore.java

```diff
@@ -95,23 +97,48 @@ public BuildRecord storeCompletedBuild(BuildRecord.Builder buildRecordBuilder, S
     }
 
     /**
-     * Match the artifacts in the list to existing artifacts in the database
-     * If no matching artifact exists in the db, the artifact is kept as-is
+     * Save the unsaved artifacts in the list to the database.  For each artifact in the given list,
+     * check if the artifact exists in the database.  If it already exists, replace it with the JPA
+     * entity loaded from the database.  If it does not exist, add it to the database.
      * 
-     * @param artifacts to store to the db
-     * @return List of artifacts stored in db
+     * @param List of in-memory artifacts to either insert to the database or find the matching record in the db
+     * @return List of up to date JPA artifact entities
      */
     private List<Artifact> saveArtifacts(List<Artifact> artifacts) {
-        List<Artifact> dbArtifacts = new ArrayList<>();
+        List<Artifact> savedArtifacts = new ArrayList<>();
         for (Artifact artifact : artifacts) {
             Artifact artifactFromDb = artifactRepository
                     .queryByPredicates(withIdentifierAndChecksum(artifact.getIdentifier(), artifact.getChecksum()));
             if (artifactFromDb == null) {
                 artifactFromDb = artifactRepository.save(artifact);
             }
-            dbArtifacts.add(artifactFromDb);
+            savedArtifacts.add(artifactFromDb);
+        }
+        return savedArtifacts;
+    }
+
+    /**
+     * Match the built artifacts in the list to existing artifacts in the database
+     * If no matching artifact exists in the db, the artifact is kept as-is
+     * 
+     * @param artifacts to store to the db
+     * @return List of artifacts stored in db
+     */
+    private List<BuiltArtifact> saveBuiltArtifacts(List<BuiltArtifact> builtArtifacts) {
+        List<BuiltArtifact> savedArtifacts = new ArrayList<>();
+        for (BuiltArtifact builtArtifact : builtArtifacts) {
+            Artifact artifactFromDb = artifactRepository
+                    .queryByPredicates(withIdentifierAndChecksum(builtArtifact.getIdentifier(), builtArtifact.getChecksum()));
+            if (artifactFromDb == null) {
+                BuiltArtifact buildArtifactFromDb = (BuiltArtifact)artifactRepository.save(builtArtifact);
+                savedArtifacts.add(buildArtifactFromDb);
+            } else if (artifactFromDb instanceof BuiltArtifact) {
+                savedArtifacts.add((BuiltArtifact)artifactFromDb);
+            } else {
+                // TODO: how should we handle when a built artifact matches an existing imported artifact?
```

Comment:
This should be an error IMO. We should not store the artifact with the same GAV.